### DISCUSSION TITLE
Is this lib better than libwebrtc?

### DISCUSSION THREAD

**Discussion by Laky-64** [paullouisageneau/libdatachannel#1162] [Q&A]
I've been having several arguments with webrtc lately (libwebrtc), like the number of connections is really very limited, you just need to get to 70 active peerconnections to send it to toomanyfilesopen, and well, reading for the network can depend on the fact that it opens many webrtc filedescriptors of its own, and I've heard absurd things, that some say it opens 200 sockets for each peer connection, does this lib also suffer from these problems?

**Comment by paullouisageneau** [comment]
libdatachannel was designed to be lightweight, it is way easier to build and resource usage is significantly lower (for instance, each peer connection typically opens a single socket). Therefore, the library can scale to hundreds or even thousands of parallel connections.

**Reply by Laky-64** [reply]
And what about the async? Is libdatachannels async? And how it handles things like "TooManyFilesOpen"? It uses boost asio?

**Reply by paullouisageneau** [reply]
libdatachannel is asynchronous, methods are non-blocking, the library handles polling by itself, and it triggers callbacks from its internal thread pool. It does not use boost asio but you can easily use it in a project based on asio (for instance you can refer to [the WebTorrent implementation in libtorrent](https://github.com/arvidn/libtorrent/blob/master/src/rtc_stream.cpp)).

"Too many files open" situation would typically translate in an exception when attempting to gather ICE candidates, with more details about the error in the log. Note that if opening a new socket for each connection is a issue, the library has a UDP multiplexing mode where peer connections share a single socket.

**Reply by pfeatherstone** [reply]
@paullouisageneau If/when using Asio, is there a way to use a working subset of this library ? This library implements thread pools, polling, sockets etc, when all of that can be handled by Asio. Is there a way to add a thin veneer of webrtc parsing on top of Asio sockets?

**Reply by pfeatherstone** [reply]
I want to avoid having two different runtimes (thread pools and epoll loops) running sockets (i'm using other connection types). Ideally Asio manages all things sockets, and this library adds parsing and webrtc state machine on top of those sockets. That should substantially reduce the amount of code required for this library to work.

**Reply by paullouisageneau** [reply]
libdatachannel was designed not to rely on any external framework for simplicity. There is no easy way to run it on top of Asio, while it is possible in theory, you'd have to refactor a lot of things in both libdatachannel and libjuice as threads are not only used to poll sockets, but also to process data, run user callbacks, etc (related issue: https://github.com/paullouisageneau/libjuice/issues/230).

**Reply by pfeatherstone** [reply]
Yeah I get that minimizing dependencies is a common thing these days but Asio is "the" canonical library for sockets in C++, it's header only and provides quite a lot out of the box. If simplicity is the goal, Asio can makes things a lot easier and simpler by not having to write a custom socket service library. I'll have a look to see how much of a refactor is required. I want to avoid several libraries using their own threadpools and threads if i can avoid it.

**Reply by pfeatherstone** [reply]
I'm writing code for a single-core computer and i can't afford all my dependencies to roll out their own thread pools and do their own socket management. But that's my tiny use case.

**Reply by paullouisageneau** [reply]
Asio being popular, easy to use, or header-only does not change that it's still a dependency to manage. Note this is not only about libdatachannel, the underlying C libraries usrsctp and libjuice will still run their own threads. FYI you can reduce the threadpool size to 2 threads here: https://github.com/paullouisageneau/libdatachannel/blob/b085ef508a2e7fcf24e8c9cde18a3b7903a77a0f/src/impl/internals.hpp#L48

**Reply by pfeatherstone** [reply]
Yeah i was wondering why you need a library for sctp ? You can create sctp posix sockets, which can then be used with Asio again. So if you remove the usrsctp library and add Asio, you're quids in. Plus by using Asio, you can remove a sh** load of code in both libdatachannel and libjuice. Just a thought.

**Reply by pfeatherstone** [reply]
If you use boost::beast, you can also remove all http and websocket code as well which would be neat. By using asio and beast you could probably remove a substantial amount of code in this library, and likely end up with less dependencies overall. I might give this a go at some point. Using a single runtime, i.e. `asio::io_context`, would be a big win for me. Then literally everything can run on the main thread which runs `asio::io_context::run()`

**Reply by pfeatherstone** [reply]
Sorry, I'm not trying to get you to rewrite your library. It's fantastic! I was just curious as to the design decisions. I can see in your BIO you're also an ASIO user so wondered why you didn't go down that route.

**Reply by paullouisageneau** [reply]
> Yeah i was wondering why you need a library for sctp ? You can create sctp posix sockets, which can then be used with Asio again. So if you remove the usrsctp library and add Asio, you're quids in. Plus by using Asio, you can remove a sh** load of code in both libdatachannel and libjuice. Just a thought.

No you can't, please stop speculating about what you don't know. You can't create SCTP sockets. First, libdatachannel runs on Windows, where they don't exist. Then, WebRTC requires an userland SCTP stack anyway because you need to tunnel SCTP over DTLS, you can't use the kernel stack. Finally, libjuice is written in C, not C++, you can't port it to Asio.

> If you use boost::beast, you can also remove all http and websocket code as well which would be neat. By using asio and beast you could probably remove a substantial amount of code in this library, and likely end up with less dependencies overall. I might give this a go at some point. Using a single runtime, i.e. `asio::io_context`, would be a big win for me. Then literally everything can run on the main thread which runs `asio::io_context::run()`

You could use beast, it's a good library. libdatachannel has a `NO_WEBSOCKET` flag to completely disable Websocket code. Just understand that not everybody has the same use case as you.

**Reply by pfeatherstone** [reply]
Yeah sorry. I didn't mean to upset anyone. Though you can do SCTP with posix. https://man7.org/linux/man-pages/man7/sctp.7.html

**Reply by paullouisageneau** [reply]
> Sorry, I'm not trying to get you to rewrite your library. It's fantastic! I was just curious as to the design decisions. I can see in your BIO you're also an ASIO user so wondered why you didn't go down that route.

Indeed, I have used boost:asio on some projects, but for libdatachannel I decided to go without it as it seemed to me the cons outweighted the pros. Besides adding a dependency, the main issue was that integrating with other libraries like usrsctp, GnuTLS, or libnice seemed like a pain.

> Yeah sorry. I didn't mean to upset anyone. Though you can do SCTP with posix. https://man7.org/linux/man-pages/man7/sctp.7.html

You can indeed do SCTP over IP (more recently SCTP over UDP) with Linux and *BSD (It is not POSIX, for instance there is no out-of-the box support on MacOS), but for WebRTC you need SCTP over DTLS with ICE, which requires an userland stack (SCTP traffic is not sent directly on the network but encapsulated in DTLS).
